src/query/storage/m3/consolidators/multi_fetch_result.go

```diff
@@ -223,33 +247,51 @@ func (r *multiResult) Add(
 			r.dedupeMap = newTagDedupeMap(opts)
 		}
 
-		r.addOrUpdateDedupeMap(r.seenFirstAttrs, first)
-		return
+		added = r.addOrUpdateDedupeMap(r.seenFirstAttrs, first)
+	} else {
+		// Now de-duplicate
+		added = r.addOrUpdateDedupeMap(attrs, newIterators)
 	}
 
-	// Now de-duplicate
-	r.addOrUpdateDedupeMap(attrs, newIterators)
+	// the series limit was reached by adding the results of this namespace to the existing results.
+	if !added {
+		r.metadata.Exhaustive = false
+		if r.limitOpts.RequireExhaustive {
+			r.err = r.err.Add(errors.ErrSeriesLimit)
+		}
+	}
 }
 
 func (r *multiResult) addOrUpdateDedupeMap(
 	attrs storagemetadata.Attributes,
 	newIterators encoding.SeriesIterators,
-) {
+) bool {
 	for _, iter := range newIterators.Iters() {
 		tagIter := iter.Tags()
 		shouldFilter, err := filterTagIterator(tagIter, r.tagOpts.Filters())
 		if err != nil {
 			r.err = r.err.Add(err)
-			return
+			return false
 		}
 
 		if shouldFilter {
 			// NB: skip here, the closer will free the series iterator regardless.
 			continue
 		}
 
-		if err := r.dedupeMap.add(iter, attrs); err != nil {
+		if r.dedupeMap.len() == r.limitOpts.Limit {
+			updated, err := r.dedupeMap.update(iter, attrs)
```

Comment:
this was rather tricky. needed to handle the case of a "better result" from a namespace overwriting an existing result.